core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java

```diff
@@ -0,0 +1,135 @@
+        return authority;
+    }
+
+    public String getHost() {
+        return host;
+    }
+
+    /**
+     * Returns the port number of this authority.
+     *
+     * @return The port component of this URI,
+     *          or {@code -1} if the port is undefined
+     */
+    public int getPort() {
+        return port;
+    }
+
+    /**
+     * Attempts to parse this URI's authority component and return {@link SchemeAndAuthority}.
+     * The authority part may have one of the following formats (The userinfo part will be ignored.):
+     * <ul>
+     *   <li>{@code "unix$3A<socket file>"} for a domain socket authority </li>
+     *   <li>{@code "<host>:<port>"} for a host endpoint </li>
+     *   <li>{@code "<host>"}, {@code "<host>:"} for a host endpoint with no port number specified</li>
+     * </ul>
+     * An IPv4 or IPv6 address can be specified in lieu of a host name, e.g. {@code "127.0.0.1:8080"} and
+     * {@code "[::1]:8080"}.
+     *
+     * @throws IllegalArgumentException if {@code authority} do not comply with RFC 2396
+     */
+    public static SchemeAndAuthority fromAuthority(String authority) {
+        requireNonNull(authority, "authority");
+
+        if (authority.startsWith("unix%3A") || authority.startsWith("unix%3a")) {
+            return new SchemeAndAuthority(null, authority, authority, -1);
+        }
+
+        final String authorityWithoutUserInfo = removeUserInfo(authority);
+        try {
+            final URI uri = new URI(null, authorityWithoutUserInfo, null, null, null).parseServerAuthority();
+            return new SchemeAndAuthority(null, uri.getRawAuthority(), uri.getHost(), uri.getPort());
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Attempts to parse this URI's authority component and return {@link SchemeAndAuthority}.
+     * The authority part may have one of the following formats (The userinfo part will be ignored.):
+     * <ul>
+     *   <li>{@code "unix$3A<socket file>"} for a domain socket authority </li>
+     *   <li>{@code "<host>:<port>"} for a host endpoint </li>
+     *   <li>{@code "<host>"}, {@code "<host>:"} for a host endpoint with no port number specified</li>
+     * </ul>
+     * An IPv4 or IPv6 address can be specified in lieu of a host name, e.g. {@code "127.0.0.1:8080"} and
+     * {@code "[::1]:8080"}.
+     *
+     * @throws IllegalArgumentException if {@code scheme} or {@code authority} do not comply with RFC 2396
+     */
+    public static SchemeAndAuthority fromSchemeAndAuthority(String scheme, String authority) {
+        requireNonNull(scheme, "scheme");
+        requireNonNull(authority, "authority");
+
+        if (!SCHEME_VALIDATOR.test(scheme)) {
+            throw new IllegalArgumentException("scheme: " + scheme + " (expected: a valid scheme)");
+        }
+        scheme = Ascii.toLowerCase(scheme);
+
+        final SchemeAndAuthority authorityObj = fromAuthority(authority);
```

Comment:
Minor suggestion What do you think of letting `scheme` be `@Nullable` and combining the two methods in interest of saving an object allocation? I don't think we need to be too strict with input parameters since this is an internal API (it is under the `*internal*` package)

e.g.
```java
... fromSchemeAndAuthority(@Nullable scheme, String authority) {
  if (scheme != null) {
    if (!SCHEME_VALIDATOR.test(scheme)) {
      ... // do scheme related validation
  }
  // do authority related parsing
  return new SchemeAndAuthority(...);
}
```

Reply:
I don't think it's good for code readability to put a parameter as null when using the method from the outside, 

ex. 
```
final SchemeAndAuthority schemeAndAuthority = SchemeAndAuthority.fromSchemeAndAuthority(null, key);
```

but if it's not good to have two object allocation methods managed, I can combining them.
**In general**, how do implement it in armeria in these cases??
provide two types of methods?? or set the parameter to nullable and provide only one method??

Reply:
> I don't think it's good for code readability to put a parameter as null when using the method from the outside 

I agree with this point

>  if it's not good to have two object allocation methods managed, I can combining them.

`SchemeAndAuthority` is not meant to be used publicly by Armeria users. (It's under the `internal` package)
For this reason, I think it's reasonable to optimize for performance at the cost of readability.

Reply:
Okay. I'll combine these methods!!